### PR TITLE
Initial monorepo build commit with Electron

### PR DESCRIPTION
**Pull request by user411**
This moves to monorepo based building for modules and moves modules to ES Modules. Also includes wallet-electron builder for Electron encapsulated native wallet builds.

** This is probably not ready to merge. Needs tested in all aspects **

I have tested building and running Suredbits Wallet and Krystal Bull locally over appServer and oracleServer and they seem to work fine.

For Electron testing, use `npm run start` to test locally for now. The native build output flashes a window for a moment and then disappears on my local. Need to debug.

### CODE REVIEW

**Comment by Christewart** [comment]
When running the `wallet-electron/npmBuild.sh` script i get this error

Screenshot from 2022-03-23 09-13-26

**Comment by user411** [comment]
Have you done an `npm i` on the base monorepo and `npm i` on wallet-electron?

**Comment by Christewart** [comment]
> Have you done an `npm i` on the base monorepo and `npm i` on wallet-electron?

That got me unblocked

**Comment by Christewart** [comment]
This is the error i got when running `npm run build` in the `oracle-electron-test` directory. As a side note, installing the `webpack-cli` automatically would be great so when we move these builds to CI, we don't require user interaction.

Screenshot from 2022-04-29 10-41-44

**Comment by Christewart** [comment]
It looks like i get a similar error when running `npm run build` inside of the `wallet-electron-ts` directory on  5398d87216e35e680a07b01a6d55d397ebef0060. I've also attached the user interaction required to install `webpack-cli` in case that is helpful

Screenshot from 2022-04-29 10-46-16
Screenshot from 2022-04-29 10-45-38

**Comment by user411** [comment]
> This is the error i got when running `npm run build` in the `oracle-electron-ts` directory on [5398d87](https://github.com/bitcoin-s/bitcoin-s-ts/commit/5398d87216e35e680a07b01a6d55d397ebef0060) . As a side note, installing the `webpack-cli` automatically would be great so when we move these builds to CI, we don't require user interaction.

I think this is linking to the published version that has `npm install` commands that are not valid. I'm looking at where "*" is used locally vs "file:../xxx" to standardize this and maybe fix the problem.

**Comment by Christewart** [comment]
So trying again on 4c3bb00370f3b53a0f6f8a1b3f6db40313012c91

I ran `npm i` in the root of the project, and the ran `npm i` in `oracle-electron-ts`. 

The next command I ran was `npm run build` and got this error

Screenshot from 2022-04-29 13-11-54

**Comment by user411** [comment]
> I ran `npm i` in the root of the project, and the ran `npm i` in `oracle-electron-ts`.

Did you do a `npm run build` after `npm i` in the root? That has potential to help...

**Comment by Christewart** [comment]
> > I ran `npm i` in the root of the project, and the ran `npm i` in `oracle-electron-ts`.
> 
> Did you do a `npm run build` after `npm i` in the root? That has potential to help...

I did not, after running `npm run build` in root, and then `npm i && npm run build && npm run make` in `oracle-electron-ts` I get this error now on  4c3bb00370f3b53a0f6f8a1b3f6db40313012c91

Screenshot from 2022-04-29 13-50-31

**Comment by user411** [comment]
> I did not, after running `npm run build` in root, and then `npm i && npm run build && npm run make` in `oracle-electron-ts` I get this error now on [4c3bb00](https://github.com/bitcoin-s/bitcoin-s-ts/commit/4c3bb00370f3b53a0f6f8a1b3f6db40313012c91)

It's weird that it's hanging up on wallet-server-proxy-ui-bundle in the oracle-electron-ts build and that it's looking for it in the root node_modules. That module is specific to wallet-electron-ts/bin.

**Comment by Christewart** [comment]
> On a fresh repo I had the same behavior trying to `npm i && npm run build && npm run make` in `oracle-electron-ts`. It went away after `npm i && npm run build && npm run make` in `wallet-electron-ts`.

The app boots up for me after building `wallet-electron-ts` as well :tada: 

Screenshot from 2022-04-29 14-46-48

**Comment by Christewart** [comment]
## Krystal Bull 

Running on 6d007cffd43a3234c13792bc21d12e015e382ffe

I do get Krystal Bull booted up with `npm run start` 🎉 
Screen Shot 2022-05-03 at 12 32 37 PM

### Issues

During `npm run build`, i see this message in my console. Since i was able to start Krystal Bull ok, maybe it isn't a big deal. I thought I would document it any way. I fetched [`bitcoin-s-oracle-server from here`](https://github.com/bitcoin-s/bitcoin-s/actions/runs/2258368162)

The electron app doesn't seem to close out correctly. I started the app with `npm run start` in development mode. When clicking the `X` to close out the GUI, the GUI disappears. However it doesn't seem I get my command line back

Screen Shot 2022-05-03 at 12 32 50 PM

I had to `ctrl+c` to kill the process and that gives me my command line back.

Screen Shot 2022-05-03 at 12 35 56 PM


It also appears that mac has some `JAVA_HOME` issues. Here is an image where I display that `JAVA_HOME` is set on the terminal, but it doesn't seem to get acknowledged by the application. This does not occur when doing `npm run start`. Perhaps this is an issue with my own development configuration.

Screen Shot 2022-05-03 at 12 49 30 PM

## Suredbits Wallet

Running on 6d007cffd43a3234c13792bc21d12e015e382ffe

I get Suredbits Wallet booted up with `npm run start` 🎉 

Screen Shot 2022-05-03 at 12 58 45 PM


I have the same issue wrt to closing out the application. When I click the `X` on the GUI it doesn't seem to kill the app when i start it with `npm run start`. I have to kill it via the command line to get the application to shtudown

Screen Shot 2022-05-03 at 12 55 56 PM

The same `JAVA_HOME` error I documented in the Krystal Bull section occurs here too

**Comment by Christewart** [comment]
> The electron app doesn't seem to close out correctly. I started the app with `npm run start` in development mode. When clicking the `X` to close out the GUI, the GUI disappears. However it doesn't seem I get my command line back

This is not an issue on linux. The app successfully exits cleanly

Screenshot from 2022-05-03 13-04-58

**Comment by Christewart** [comment]
On linux, I ran `npm run make` to produce a `deb`. I installed the deb with `sudo dpkg -i krystal-bull_0.0.1_amd64.deb` 

When attempting to start the application built, I get these error messages in my `/var/log/syslog` and the app crashes

Screenshot from 2022-05-03 13-09-28

**Comment by Christewart** [comment]
It appears the wallet-electron deb installation has the same problem as Krystal Bull. Here is the output in `/var/log/syslog`

Screenshot from 2022-05-03 13-24-45

**Comment by user411** [comment]
I put in a commit to turn off file logging from the electron apps, which were the build errors you saw.

Mac applications don't close by default when the window is closed. You need to Quit (Cmd-Q) the application.

There is documentation on setting JAVA_HOME on a Mac in the README.md on the electron apps. You need to

```
launchctl setenv JAVA_HOME $JAVA_HOME
```

in ~/.zprofile

**Comment by user411** [comment]
Might be worth testing the following in webpack.main.config.js:

```
new PermissionsOutputPlugin({
      buildFiles: [{
        path: path.resolve(packagedPath, `bin/${server}/bin/${server}`),
        fileMode: '755'
      }, {
        path: path.resolve(packagedPath, `bin/${server}/bin/${server}.bat`),
        fileMode: '755'
      }, {
        path: path.resolve(packagedPath, `bin/${server}/jre/bin/java`),
        fileMode: '755'
      }, {
        path: path.resolve(packagedPath, `bin/${server}/jre/bin/keytool`),
        fileMode: '755'
      }
    ],
```

**Comment by user411** [comment]
The `mv` line is not a terminal error, just based on zip naming not having extra fluff. It doesn't negatively impact anything in my experience. Feel free to make the * name matching in the .sh better if you know a better way to account for the naming variation on some zips with build info.

**Comment by Christewart** [comment]
Ok, so I'm going to attempt to describe the behavior i'm seeing. 

1. I do `npm i && npm run build` in the root of the project
2. I do `npm i && npm run build` in `oracle-electron-ts`
3. I run `npm run start`

I get different behavior each time, ~50% of the time it works, and then ~50% of the time I can get some flavor of https://github.com/bitcoin-s/bitcoin-s/issues/4325

Since its not consistent, I do wonder if there is some sort of issue on the `bitcoin-s-ts` side rather than the artifact packaging

Here is an example of it working:
Screenshot from 2022-05-09 12-39-00

Here are some various failures where different `ClassNotFoundException`

Screenshot from 2022-05-09 12-38-03
Screenshot from 2022-05-09 12-37-39
Screenshot from 2022-05-09 12-33-43

File: wallet-electron-ts/makeBin.sh [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+cp -a ../wallet-server-ui-proxy/dist/cjs/build.json bin/wallet-server-ui-proxy
+cp -a ../wallet-server-ui-proxy/dist/bundle-static.js bin/wallet-server-ui-proxy
+# Replace config.json
+cp -a proxy-config.json bin/wallet-server-ui-proxy/config.json
+
+# Expand bitcoin-s-server zip
+echo "TODO : Pull down bitcoin-s-server.zip"
+mkdir -p bin/bitcoin-s-server
+unzip bitcoin-s-server.zip -d bin/bitcoin-s-server
+chmod +x bin/bitcoin-s-server/bin/bitcoin-s-server
```
Comment (Christewart) [inline wallet-electron-ts/makeBin.sh:38]: I think this is an accidental duplicate `bin/bitcoin-s-server/bin/bitcoin-s-server`

Screenshot from 2022-05-09 13-11-57

Comment (user411) [reply wallet-electron-ts/makeBin.sh:38]: How do you mean? Where does it live? If you mean the one in .webpack/, that is the final bin folder that is built for the Electron app.

File: wallet-electron-ts/webpack.main.config.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      patterns: [
+        { from: path.resolve(__dirname, 'assets'), to: 'assets' },
+        // This is destroying +x permission on items in bin
+        { from: path.resolve(__dirname, 'bin'), to: 'bin' },
+        { from: path.resolve(__dirname, 'proxy-config.json'), to: 'config.json' },
+        { from: path.resolve(__dirname, 'bin', 'wallet-server-ui-proxy', 'build.json'), to: 'build.json' },
+      ]
+    }),
+    // Re-set the +x permissions on bitcoin-s-server executables
+    new PermissionsOutputPlugin({
```
[the remark this answers is not in the document]
Comment (user411) [reply wallet-electron-ts/webpack.main.config.js:35]: Awesome, expected that would be required.

**Comment by Christewart** [comment]
> I'm going to see if this behavior is reproducible on `wallet-electron-ts`

I was not able to reproduce any of this behavior on `wallet-electron-ts` :thinking: .

**Comment by Christewart** [comment]
> > I'm going to see if this behavior is reproducible on `wallet-electron-ts`
> 
> I was not able to reproduce any of this behavior on `wallet-electron-ts` thinking .

I can't even reproduce this on `oracle-electron-ts` now. I'm unsure of how I got to this state to begin with :thinking:

**Comment by Christewart** [comment]
I believe the `JAVA_HOME` logic needs to be removed now that we ship with a jre. When trying to install off of mac i get this error on a19aac11e334a12310f11ed3903f134e13d62e80
 
Screen Shot 2022-05-09 at 6 34 22 PM

**Comment by Christewart** [comment]
It looks like the `-mac.zip` suffix on `bitcoin-s-server-mac.zip` is giving the script a problem when attempting to build the wallet server backend

Screen Shot 2022-05-09 at 6 46 31 PM

File: wallet-electron-ts/makeBin.sh (hunk 1 above)
Comment (Christewart) [reply wallet-electron-ts/makeBin.sh:38]: I guess it seems to me that `bin/bitcoin-s-server/bin/bitcoin-s-server` is redundant, but I think the root problem was fixed in a19aac11e334a12310f11ed3903f134e13d62e80 so this is just a naming issue. I thought there was something deeper, i'm resolving this thread for now

**Pull request merged**